File: common/go_common/mcached_manipulate.go
// ----------------------------------------------------------------
//
//	mcached_manipulate.go
//
//					May/20/2018
//
// ----------------------------------------------------------------
package main
 
import (
	"fmt"
	"net"
	"strconv"
	"strings"
	"encoding/json"
)

// ----------------------------------------------------------------
func socket_read_proc (conn net.Conn,key_in string) string {
	str_received := ""
	_, err := conn.Write([]byte("get " + key_in + "\r\n"))
	if err != nil {
		fmt.Println(err)
		return str_received
		}

	buf := make([]byte, 1024)
	nn, err := conn.Read(buf[:])
	if err != nil {
		fmt.Println(err)
		return str_received
		}

	str_received = string(buf[0:nn])

	return	str_received
}

// ----------------------------------------------------------------
func mcached_socket_read_proc (conn net.Conn,key_in string) string {
	json_str := ""
	str_received := socket_read_proc (conn,key_in)

	lines := strings.Split(str_received,"\n")

	if (! strings.Contains(lines[0],"END")) {
		json_str = lines[1]
		}

	return	json_str
}

// ----------------------------------------------------------------
func json_parser (key string,json_str string) {
//			fmt.Println (json_str)
	var unit_aa map[string]interface{}

	json.Unmarshal ([]byte(json_str), &unit_aa)
	fmt.Printf ("%s\t",key)
	fmt.Printf ("%s\t",unit_aa["name"])
	fmt.Printf ("%f\t",unit_aa["population"])
	fmt.Printf ("%s\n",unit_aa["date_mod"])
}

// ----------------------------------------------------------------
func mcached_socket_write_proc (conn net.Conn,key_in string,json_str string) {
//	fmt.Println ("*** mcached_socket_write_proc ***")

	llx := len (json_str)
//	fmt.Println (llx)

	llx_str  := strconv.Itoa (llx)

	comm_aa := "set " + key_in + " 0 0 " + llx_str + "\r\n"
	conn.Write([]byte(comm_aa))

	comm_bb := json_str + "\r\n"
	conn.Write([]byte(comm_bb))

	buf := make ([]byte,1024)
	conn.Read (buf[:])

//	nn,_ := conn.Read(buf[:])
//	fmt.Println (string(buf[0:nn]))
}

// ----------------------------------------------------------------
func json_update_proc (json_str string,population_in int) string {

	var unit_aa map[string]interface{}
	json.Unmarshal ([]byte(json_str), &unit_aa)
	fmt.Printf ("%s\t",unit_aa["name"])
	fmt.Printf ("%f\t",unit_aa["population"])
	fmt.Printf ("%s\n",unit_aa["date_mod"])

	unit_aa["population"] = population_in
	unit_aa["date_mod"] = get_current_date_proc ()

	output, _ := json.Marshal(unit_aa)

	json_str = string(output)

	return	json_str
}

// ----------------------------------------------------------------
func dict_to_mcached_proc (hostname string,port int,dict_aa map[string](map[string]interface{})) {
	str_port := strconv.Itoa (port)
	conn, err := net.Dial ("tcp",hostname + ":" + str_port)
	if err != nil {
		fmt.Println(err)
		return
		}

	for key,value := range dict_aa {
//		fmt.Print (key + "\t")
//		fmt.Println (value["name"])
		output, _ := json.Marshal (value)

		json_str := string(output)
//		fmt.Println (json_str)
		mcached_socket_write_proc (conn,key,json_str)
		}

	conn.Close ()
}

// ----------------------------------------------------------------
func mcached_to_dict_proc (hostname string,port int,keys []string) map[string](map[string]interface{}){
	dict_aa := make (map[string](map[string]interface{}))

	str_port := strconv.Itoa (port)
	conn, err := net.Dial ("tcp",hostname + ":" + str_port)
	if err != nil {
		fmt.Println(err)
		return dict_aa
		}

	for _,key := range keys {
		json_str := mcached_socket_read_proc (conn,key)

		if 0 < len (json_str) {
			var unit_aa map[string]interface{}
			json.Unmarshal ([]byte(json_str), &unit_aa)
			dict_aa[key] = unit_aa
			}
	}
	conn.Close ()

	return (dict_aa)
}

// ----------------------------------------------------------------
func mcached_update_proc (hostname string,port int,key_in string,population_in int) {
	str_port := strconv.Itoa (port)
	conn, err := net.Dial ("tcp",hostname + ":" + str_port)
	if err != nil {
		fmt.Println(err)
		return
		}

	json_str := mcached_socket_read_proc (conn,key_in)

	if 0 < len (json_str) {
		json_str_new := json_update_proc (json_str,population_in)
		fmt.Println (json_str_new)
		mcached_socket_write_proc (conn,key_in,json_str_new)
		}

	conn.Close ()
}

// ----------------------------------------------------------------
func mcached_delete_proc (hostname string,port int,key_in string) {
	str_port := strconv.Itoa (port)
	conn, err := net.Dial ("tcp",hostname + ":" + str_port)
	if err != nil {
		fmt.Println(err)
		return
		}

	_, err = conn.Write([]byte("delete " + key_in + "\r\n"))
	if err != nil {
		fmt.Println(err)
		return
		}

	conn.Close ()
}

// ----------------------------------------------------------------

File: common/lua_common/mcached_manipulate.lua
-- ----------------------------------------------------
--	mcached_manipulate.lua
--
--					Feb/16/2015
-- ----------------------------------------------------
json=require	('json')
-- ----------------------------------------------------
function mcached_get_value_proc (client,key)
	local obj = nil
	local command = "get " .. key .. "\r\n"
	client:send(command)
	local ret_0 = client:receive()
	if (ret_0 ~= "END") then
		local ret_1 = client:receive()
--		print( "response: " .. ret_1 )
		obj = json.decode(ret_1)
		local ret_2 = client:receive()
	end

	return	obj
end

-- ----------------------------------------------------
function mcached_to_dict_proc (server_host,server_port,keys)
	dict_aa = {}
	local socket = require('socket')

	local client = socket.connect (server_host,server_port)


	for it in pairs (keys)  do
		local key = keys[it]

		local obj = mcached_get_value_proc (client,key)
		if (obj) then
			dict_aa[key] = obj
		end
	end

	client:close()

	return	dict_aa
end

-- ----------------------------------------------------
function mcached_delete_proc (server_host,server_port,key_in)
	local socket = require('socket')

	local client = socket.connect (server_host,server_port)

	local command = "delete " .. key_in .. "\r\n"
	client:send(command)
	local ret_0 = client:receive()
	print (ret_0)

	client:close()
end

-- ----------------------------------------------------
function mcached_set_value_proc (client,key,unit_aa)
	str_json = json.encode (unit_aa)
	llx =  string.len (str_json)

	command = "set " .. key .. " 0 0 " .. llx .. "\r\n"
	client:send(command)
	client:send(str_json .. "\r\n")
	ret_0 = client:receive()
	print (ret_0)
	
end
-- ----------------------------------------------------
function mcached_update_proc (server_host,server_port,key_in,population_in)
	local socket = require('socket')
	local client = socket.connect (server_host,server_port)

	local unit_aa = mcached_get_value_proc (client,key_in)
	if (unit_aa) then
		unit_aa.population=population_in
		unit_aa.date_mod=get_current_date_proc ()

		mcached_set_value_proc (client,key_in,unit_aa)

	end
	client:close ()
end

-- ----------------------------------------------------
function dict_to_mcached_proc (server_host,server_port,dict_aa)

	local socket = require('socket')

	local client = socket.connect (server_host,server_port)

	for key,unit_aa in pairs (dict_aa)
		do
			mcached_set_value_proc (client,key,unit_aa)
		end

	client:close()
end

-- ----------------------------------------------------
